README.md
< [English version here](README_en.md) >

[![Travis bormeparser](https://travis-ci.org/PabloCastellano/bormeparser.svg?branch=master)](https://travis-ci.org/PabloCastellano/bormeparser)
[![Pypi bormeparser](https://badge.fury.io/py/bormeparser.png)]( https://pypi.python.org/pypi/bormeparser)
[![Downloads bormeparser](https://img.shields.io/pypi/dm/bormeparser.svg)](https://pypi.python.org/pypi/bormeparser)
[![Coverage Status](https://coveralls.io/repos/PabloCastellano/bormeparser/badge.svg)](https://coveralls.io/r/PabloCastellano/bormeparser)
[![Documentation Status](https://readthedocs.org/projects/bormeparser/badge/?version=latest)](https://readthedocs.org/projects/bormeparser/?badge=latest)

**NOTA: Este repositorio se encuentra desmantenido. Si buscas una solución para acceder a la información mercantil mediante API y con soporte para empresas, consulta [LibreBOR](https://librebor.me) y [Documentación LibreBOR API](https://docs.librebor.me/).**

bormeparser
===========

**bormeparser** es una librería de **Python 3** para parsear los archivos del BORME (Boletín Oficial del Registro Mercantil en España).

Qué es BORME
============

El **Boletín Oficial del Registro Mercantil** es un documento publicado diariamente por el Registro Mercantil Central (RMC)
en España que contiene un listado de las últimas sociedades creadas y disueltas así como otros datos que las empresas
están obligadas a comunicar.

La librería aprovecha que desde la aprobación de [esta ley](https://www.boe.es/buscar/doc.php?id=BOE-A-2008-19826),
desde el año 2009 el BORME se publica también en formato electrónico con la misma validez que su versión en papel.

Los BORMEs se publican en https://boe.es/diario_borme/.

Desgraciadamente debido al acuerdo actual con el Registro Mercantil, no pueden publicar todos los datos en un formato
útil y reutilizable como XML o JSON y los datos más interesantes están solo disponibles en los archivos PDF.

Puedes leer más sobre ello en:
- Wikipedia: https://es.wikipedia.org/wiki/Boletín_Oficial_del_Registro_Mercantil


Documentación y uso
===================

El directorio docs/ contiene toda la documentación. Puedes generarla ejecutando:

    cd docs && make html

Si quieres generarla para otro idioma (inglés en este caso):

    make -e SPHINXOPTS="-D language='en'" html

Puedes consultar la versión online en http://bormeparser.readthedocs.org/es/latest/.


Instalación desde Git
=====================

Puedes instalar bormeparser descargándolo y ejecutando:

    sudo apt-get install python3-dev libxslt1-dev libffi-dev zlib1g-dev gcc
    git clone https://github.com/PabloCastellano/bormeparser.git
    cd bormeparser
    python setup.py install


Instalación desde Pip
=====================

    pip install bormeparser


Scripts
=======

La carpeta scripts/ contiene algunos scripts útiles para tratar archivos BORME. Todos tienen parámetros de entrada similares.

    python scripts/download_borme_pdfs.py -f init -p VALENCIA
    python scripts/check_bormes.py -f init -p VALENCIA
    python scripts/download_borme_pdfs.py -d /tmp/bormemadrid -p MADRID -f 2016-06-01 -t 2016-06-30
    python scripts/check_bormes.py -d /tmp/bormemadrid -p MADRID -f 2016-06-01 -t 2016-06-30
    python scripts/borme_json_all.py -d /tmp/bormemadrid

Tests
=====

Este paquete contiene tests unitarios. Puedes ejecutarlos escribiendo alguna de estas órdenes:

    python setup.py test
    python -m unittest bormeparser.tests.test_borme
    python -m unittest bormeparser.tests.test_bormeparser
    python -m unittest bormeparser.tests.test_bormeregex


Licencia
========

Todo el código está bajo licencia GPLv3+. Para más información consulta el archivo LICENSE.txt


CHANGES.rst
Changelog for bormeparser
=========================

0.5.2 (unreleased)
------------------

- New force download flag in download_url()


0.5.1 (unreleased)
------------------

- Nothing changed yet.


0.5.0 (2022-09-27)
------------------

- Actualizadas las dependencias
- Bormeparser requiere Python 3.6+ ahora
- Añadido Dockerfile


0.4.2 (2021-03-01)
------------------

- Actualizadas las dependencias


0.4.1 (2021-02-15)
------------------

- No generar universal wheel (py2 ya no está soportado)
- Detectar cargo "AUDIT.INDIV."
- Pequeñas mejoras en scripts check_bormes.py y download_borme_pdf.py
- Actualizada dependencia lxml (CVE-2020-27783)

0.4.0 (2019-09-18)
------------------

- Bump requirements
- Require python 3.5


0.3.5 (2019-02-18)
------------------

- Mejora en la detección del encoding de BORME-XML
- No fallar cuando la cabecera Content-Length no esté presente
- Pequeños cambios en los niveles de logging
- download: ajustados los valores por defecto de sleep y threads
- Borrados los checks de python2


0.3.4 (2018-09-24)
------------------

- Nuevo método BormeXML.get_cve_url()


0.3.3 (2018-09-23)
------------------

- is_company(): llama a clean_empresa() y comprueba si contiene la palabra "SOCIEDAD"


0.3.2 (2018-09-23)
------------------

- Actualización de las dependencias
- Tests arreglados
- Nuevo acto mercantil: Adaptación Ley 44/2015
- download_url() ahora reintenta la descarga si hubo un error
- BormeXML.save_to_file() crea el directorio si no existe
- clean_empresa(): quita "EN LIQUIDACION" y "SUCURSAL EN ESPAÑA" del nombre


0.3.1 (2018-05-29)
------------------

- Añadidos más tipos de sociedad
- Borme.from_json: permitir un objeto file como argumento filename


0.3.0 (2018-03-12)
------------------

- Eliminado soporte de Python 2
- Cambios en el formato BORME-JSON
- Nombres de actos repetidos en el mismo anuncio (issue #3)
- Usar requests en lugar de urllib
- Archivo de configuración ~/.bormecfg
- Mejoras en el parser
- Añadidos 4 nuevos actos y 41 cargos directivos
- Borme.to_json ahora permite especificar un path (archivo o directorio) en lugar de solo archivo
- Borme._set_url evita conexión a Internet si existe BORME-XML
- Sociedades y registros tienen su propio módulo
- Funciones de limpieza de datos en bormeparser.clean
- Incluye nombre del R.M. en BORME-JSON
- Cambios menores en los scripts
- Borme.XML devuelve str en lugar de list si solo hay un elemento
- BormeXML.get_provincias

0.2.4 (2016-09-21)
------------------

- BormeXML: get_url_pdfs, get_cves y get_sizes ahora permiten especificar sección y provincia
- Nueva constante ALL_PROVINCIAS en bormeparser.provincia
- Detección de nuevos tipos de sociedades
- Scripts: limpieza, uso de argparse en los scripts, unificación de parámetros
- Mejoras menores en la documentación
- Nuevos campos "version" y "raw_version" en el formato JSON de BORME

0.2.3 (2016-04-26)
------------------

- Mejora en el parser de BORME C

0.2.2 (2016-04-26)
------------------

- Mejoras en el parser de BORME C

0.2.1 (2016-04-25)
------------------

- Corregidos fallos de compatibilidad con Python 2

0.2 (2016-04-25)
----------------

- Eliminado primer argumento "date" de BormeXML.get_url_pdfs()
- Nuevo método: BormeXML.get_urls_cve()
- Arregladas algunas incompatibilidades con Python 2
- Nuevas funciones: get_borme_website(), get_url_seccion_c()
- Se incorpora parser para BORME C
- Añadido el acto "(Primera inscripcion O.M. 10/6/1.997)"
- Renombradas constantes y funciones
- BormeXML: BORME C
- script download_borme_pdfs_C.py
- Mejora parsing de cargos repetidos en el mismo acto (issue #4)

0.1.5 (2015-09-25)
------------------

- Añadidos nuevos cargos
- Mejoras en setuptools

0.1.4 (2015-09-24)
------------------

- Grandes mejoras en el parser en general
- Añadidos cargos y actos nuevos
- Mejoras en las expresiones regulares
- Los objetos Provincia ahora son comparables
- download_pdfs() ahora admite los parámetros seccion y provincia
- Nuevos scripts: borme_to_json, download_borme_pdfs_A, borme_sort, xml_poller
- Uso de OrderedDict en lugar de dict
- Uso de OrderedDict en lugar de dict
- Usar la librería logging
- Más tests
- Actualización de los requisitos

0.1.3 (2015-08-08)
------------------

- Fixed missing packages that weren't distributed

0.1.2 (2015-08-07)
------------------

- Fixed UnicodeWarning that caused tests to fail in Python 2

0.1.1 (2015-08-07)
------------------

- setup.py install now installs requirements

0.1 (2015-08-07)
----------------

- First release
- Download and parse BORME PDF files
- Main parser is PyPDF2
- Python 2 and 3 support
- Tests suite


bormeparser/acto.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# acto.py -
# Copyright (C) 2015-2022 Pablo Castellano <pablo@anche.no>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


class ACTO:
    NOMBRAMIENTOS = 1
    REVOCACIONES = 2
    CESES_DIMISIONES = 3
    MODIFICACIONES_ESTATUTARIAS = 4
    CAMBIO_DE_OBJETO_SOCIAL = 5
    CAMBIO_DE_DENOMINACION_SOCIAL = 6
    CAMBIO_DE_DOMICILIO_SOCIAL = 7
    AMPLIACION_DEL_OBJETO_SOCIAL = 8
    SOCIEDAD_UNIPERSONAL = 9
    DISOLUCION = 10
    REELECCIONES = 11
    CONSTITUCION = 12
    ARTICULO_378_5_DEL_RRM = 13
    OTROS_CONCEPTOS = 14
    AMPLIACION_DE_CAPITAL = 15
    REDUCCION_DE_CAPITAL = 16
    SITUACION_CONCURSAL = 17
    FUSION_POR_ABSORCION = 18
    SUSPENSION_DE_PAGOS = 19
    TRANSFORMACION_DE_SOCIEDAD = 20
    CANCELACIONES_DE_OFICIO_DE_NOMBRAMIENTOS = 21
    DESEMBOLSO_DE_DIVIDENDOS_PASIVOS = 22
    PAGINA_WEB_DE_LA_SOCIEDAD = 23
    PRIMERA_SUCURSAL_DE_SOCIEDAD_EXTRANJERA = 24

    EXTINCION = 26
    DECLARACION_DE_UNIPERSONALIDAD = 27
    PERDIDA_DEL_CARACTER_DE_UNIPERSONALIDAD = 28
    REAPERTURA_HOJA_REGISTRAL = 29
    ADAPTACION_LEY_2_95 = 30
    CIERRE_PROVISIONAL_BAJA_EN_EL_INDICE_DE_ENTIDADES_JURIDICAS = 31
    CIERRE_PROVISIONAL_REVOCACION_NIF = 32
    REACTIVACION_DE_LA_SOCIEDAD = 33
    FE_DE_ERRATAS = 34
    DATOS_REGISTRALES = 35
    CREDITO_INCOBRABLE = 36
    EMPRESARIO_INDIVIDUAL = 37
    EMISION_OBLIGACIONES = 38
    MODIFICACION_PODERES = 39
    ESCISION_PARCIAL = 40
    ESCISION_TOTAL = 41
    FUSION_UNION = 42
    ADAPTACION_DE_LA_SOCIEDAD = 43
    QUIEBRA = 44
    SUCURSAL = 45
    CESION_GLOBAL_ACTIVO_PASIVO = 46
    SEGREGACION = 47
    ACUERDO_AMPLIACION_CAPITAL_SOCIAL_SIN_EJECUTAR = 48
    MODIFICACION_DE_DURACION = 49
    APERTURA_DE_SUCURSAL = 50
    CIERRE_PROVISIONAL_IMPUESTO_SOCIEDADES = 51
    PRIMERA_INSCRIPCION = 52
    ANOTACION_PREVENTIVA_DEMANDA = 53
    ANOTACION_PREVENTIVA_DECLARACION = 54
    CIERRE_SUCURSAL = 55
    ADAPTACION_LEY_44_2015 = 56
    BENEFICIARIA = 57

    # Palabras clave con argumentos
    _arg_keywords = {
        'Nombramientos': NOMBRAMIENTOS,
        'Revocaciones': REVOCACIONES,
        'Ceses/Dimisiones': CESES_DIMISIONES,
        'Modificaciones estatutarias': MODIFICACIONES_ESTATUTARIAS,
        'Cambio de objeto social': CAMBIO_DE_OBJETO_SOCIAL,
        'Cambio de denominación social': CAMBIO_DE_DENOMINACION_SOCIAL,
        'Cambio de domicilio social': CAMBIO_DE_DOMICILIO_SOCIAL,
        'Ampliacion del objeto social': AMPLIACION_DEL_OBJETO_SOCIAL,
        'Disolución': DISOLUCION,
        'Reelecciones': REELECCIONES,
        'Constitución': CONSTITUCION,
        'Apertura de sucursal': APERTURA_DE_SUCURSAL,
        'Empresario Individual': EMPRESARIO_INDIVIDUAL,
        'Articulo 378.5 del Reglamento del Registro Mercantil': ARTICULO_378_5_DEL_RRM,
        'Otros conceptos': OTROS_CONCEPTOS,
        'Ampliación de capital': AMPLIACION_DE_CAPITAL,
        'Reducción de capital': REDUCCION_DE_CAPITAL,
        'Situación concursal': SITUACION_CONCURSAL,
        'Fusión por absorción': FUSION_POR_ABSORCION,
        'Absorbidas': FUSION_POR_ABSORCION,
        'Suspensión de pagos': SUSPENSION_DE_PAGOS,
        'Transformación de sociedad': TRANSFORMACION_DE_SOCIEDAD,
        'Cancelaciones de oficio de nombramientos': CANCELACIONES_DE_OFICIO_DE_NOMBRAMIENTOS,
        'Desembolso de dividendos pasivos': DESEMBOLSO_DE_DIVIDENDOS_PASIVOS,
        'Página web de la sociedad': PAGINA_WEB_DE_LA_SOCIEDAD,
        'Primera sucursal de sociedad extranjera': PRIMERA_SUCURSAL_DE_SOCIEDAD_EXTRANJERA,
        'Emisión de obligaciones': EMISION_OBLIGACIONES,
        'Modificación de poderes': MODIFICACION_PODERES,
        'Escisión parcial': ESCISION_PARCIAL,
        'Fusión por unión': FUSION_UNION,
        'Beneficiarias': BENEFICIARIA,
        'Beneficiarios de la cesion': BENEFICIARIA,
        'Quiebra': QUIEBRA,
        'Sucursal': SUCURSAL,
        'Cesión global de activo y pasivo': CESION_GLOBAL_ACTIVO_PASIVO,
        'Segregación': SEGREGACION,
        'Primera inscripcion (O.M. 10/6/1.997)': PRIMERA_INSCRIPCION,
        'Anotación preventiva. Demanda de impugnación de acuerdos sociales': ANOTACION_PREVENTIVA_DEMANDA,
        'Anotación preventiva. Declaración de deudor fallido': ANOTACION_PREVENTIVA_DECLARACION,
    }

    # Palabras clave sin argumentos
    _noarg_keywords = {
        'Crédito incobrable': CREDITO_INCOBRABLE,
        'Sociedad unipersonal': SOCIEDAD_UNIPERSONAL,
        'Extinción': EXTINCION,
        'Pérdida del caracter de unipersonalidad': PERDIDA_DEL_CARACTER_DE_UNIPERSONALIDAD,
        'Reapertura hoja registral': REAPERTURA_HOJA_REGISTRAL,
        'Adaptación Ley 2/95': ADAPTACION_LEY_2_95,
        'Adaptación Ley 44/2015': ADAPTACION_LEY_44_2015,
        'Adaptada segun D.T. 2 apartado 2 Ley 2/95': ADAPTACION_LEY_2_95,
        'Cierre provisional hoja registral por baja en el índice de Entidades Jurídicas': CIERRE_PROVISIONAL_BAJA_EN_EL_INDICE_DE_ENTIDADES_JURIDICAS,
        'Cierre provisional de la hoja registral por revocación del NIF': CIERRE_PROVISIONAL_REVOCACION_NIF,
        'Cierre provisional hoja registral por revocación del NIFde Entidades Jurídicas': CIERRE_PROVISIONAL_REVOCACION_NIF,
        'Cierre provisional hoja registral art. 137.2 Ley 43/1995 Impuesto de Sociedades': CIERRE_PROVISIONAL_IMPUESTO_SOCIEDADES,
        'Reactivación de la sociedad (Art. 242 del Reglamento del Registro Mercantil)': REACTIVACION_DE_LA_SOCIEDAD,
        'Adaptación de sociedad': ADAPTACION_DE_LA_SOCIEDAD,
        'Cierre de Sucursal': CIERRE_SUCURSAL,
    }

    # Palabras clave seguidas por :
    _colon_keywords = {
        'Modificación de duración': MODIFICACION_DE_DURACION,
        'Fe de erratas': FE_DE_ERRATAS,
    }

    _bold_keywords = {
        'Declaración de unipersonalidad': DECLARACION_DE_UNIPERSONALIDAD,
        'Sociedad unipersonal': SOCIEDAD_UNIPERSONAL,
        'Acuerdo de ampliación de capital social sin ejecutar. Importe del acuerdo': ACUERDO_AMPLIACION_CAPITAL_SOCIAL_SIN_EJECUTAR,
        'Escisión total': ESCISION_TOTAL,
    }

    # Palabra clave
    _ending_keywords = {
        'Datos registrales': DATOS_REGISTRALES
    }

    ARG_KEYWORDS = list(_arg_keywords.keys())
    NOARG_KEYWORDS = list(_noarg_keywords.keys())
    COLON_KEYWORDS = list(_colon_keywords.keys())
    BOLD_KEYWORDS = list(_bold_keywords.keys())
    ENDING_KEYWORDS = list(_ending_keywords.keys())
    ALL_KEYWORDS = ARG_KEYWORDS + NOARG_KEYWORDS + COLON_KEYWORDS + BOLD_KEYWORDS + ENDING_KEYWORDS

"""
    DICT_KEYWORDS = {kw: remove_accents(kw).replace(' del ', ' ').replace(' por ', ' ').replace(' de ', ' ')
                 .replace(' ', '_').replace('/', '_').replace('.', '_').lower() for kw in ALL_KEYWORDS}

>>> DICT_KEYWORDS.values()
['revocaciones', 'cambio_objeto_social', 'reelecciones', 'otros_conceptos', 'fe_erratas', 'sociedad_unipersonal', 'declaracion_unipersonalidad', 'constitucion', 'suspension_pagos', '
perdida_caracter_unipersonalidad', 'cancelaciones_oficio_nombramientos', 'datos_registrales', 'cambio_domicilio_social', 'disolucion', 'ampliacion_objeto_social', 'cierre_provisional_hoj
a_registral_baja_en_el_indice_entidades_juridicas', 'ceses_dimisiones', 'nombramientos', 'situacion_concursal', 'modificaciones_estatutarias', 'ampliacion_capital', 'adaptacion_ley_2_95'
, 'cambio_denominacion_social', 'extincion', 'reduccion_capital', 'cambio_identidad_socio_unico', 'transformacion_sociedad', 'reapertura_hoja_registral', 'socio_unico', 'articulo_378_5
_reglamento_registro_mercantil', 'fusion_absorcion']
"""
